src/module/iotjs_module_httpparser.c

```diff
@@ -0,0 +1,554 @@
+HTTP_CB( on_message_complete );
+
+
+const struct http_parser_settings settings = {
+  on_message_begin,
+  on_url,
+  on_status,
+  on_header_field,
+  on_header_value,
+  on_headers_complete,
+  on_body,
+  on_message_complete,
+};
+
+
+static iotjs_httpparserwrap_t* get_parser_wrap(const iotjs_jval_t* jparser) {
+  uintptr_t handle = iotjs_jval_get_object_native_handle(jparser);
+  return (iotjs_httpparserwrap_t*)(handle);
+}
+
+
+JHANDLER_FUNCTION(Reinitialize) {
+  JHANDLER_CHECK_THIS(object);
+  JHANDLER_CHECK_ARGS(1, number);
+
+  const iotjs_jval_t* jparser = JHANDLER_GET_THIS(object);
+
+  http_parser_type httpparser_type =
+    (http_parser_type)(JHANDLER_GET_ARG(0, number));
+  IOTJS_ASSERT(httpparser_type == HTTP_REQUEST ||
+               httpparser_type == HTTP_RESPONSE);
+
+  iotjs_httpparserwrap_t* parser = get_parser_wrap(jparser);
+  iotjs_httpparserwrap_initialize(parser, httpparser_type);
+}
+
+
+JHANDLER_FUNCTION(Finish) {
+  JHANDLER_CHECK_THIS(object);
+  JHANDLER_CHECK_ARGS(0);
+
+  const iotjs_jval_t* jparser = JHANDLER_GET_THIS(object);
+  iotjs_httpparserwrap_t* parser = get_parser_wrap(jparser);
+
+  http_parser* nativeparser = iotjs_httpparserwrap_parser(parser);
+  int rv = http_parser_execute(nativeparser, &settings, NULL, 0);
+
+  if (rv != 0) {
+    enum http_errno err = HTTP_PARSER_ERRNO(nativeparser);
+
+    iotjs_jval_t eobj = iotjs_jval_create_error("Parse Error");
+    iotjs_jval_set_property_number(&eobj, "byteParsed", 0);
+    iotjs_jval_set_property_string_raw(&eobj, "code", http_errno_name(err));
+    iotjs_jhandler_return_jval(jhandler, &eobj);
+    iotjs_jval_destroy(&eobj);
+  }
+}
+
+
+JHANDLER_FUNCTION(Execute) {
+  JHANDLER_CHECK_THIS(object);
+  JHANDLER_CHECK_ARGS(1, object);
+
+  const iotjs_jval_t* jparser = JHANDLER_GET_THIS(object);
+  iotjs_httpparserwrap_t* parser = get_parser_wrap(jparser);
+
+
+  const iotjs_jval_t* jbuffer = JHANDLER_GET_ARG(0, object);
+  iotjs_bufferwrap_t* buffer = iotjs_bufferwrap_from_jbuffer(jbuffer);
+  char* buf_data = iotjs_bufferwrap_buffer(buffer);
+  int buf_len = iotjs_bufferwrap_length(buffer);
+  JHANDLER_CHECK(buf_data != NULL);
+  JHANDLER_CHECK(buf_len > 0);
+
+  iotjs_httpparserwrap_set_buf(parser, (iotjs_jval_t*)jbuffer, buf_data,
+                               buf_len);
+
+  http_parser* nativeparser = iotjs_httpparserwrap_parser(parser);
+  int nparsed = http_parser_execute(nativeparser, &settings, buf_data, buf_len);
+
+  iotjs_httpparserwrap_set_buf(parser, NULL, NULL, 0);
+
+
+  if (!nativeparser->upgrade && nparsed != buf_len) {
+    // nparsed should equal to buf_len except UPGRADE protocol
+    enum http_errno err = HTTP_PARSER_ERRNO(nativeparser);
+    iotjs_jval_t eobj = iotjs_jval_create_error("Parse Error");
+    iotjs_jval_set_property_number(&eobj, "byteParsed", 0);
+    iotjs_jval_set_property_string_raw(&eobj, "code", http_errno_name(err));
+    iotjs_jhandler_return_jval(jhandler, &eobj);
+    iotjs_jval_destroy(&eobj);
+  }
+  else{
```

Comment:
Please check coding style in this line. - ``The else goes on the same line as the closing brace.``